feat: yellow paper cross-chain communication (#3477)
AztecProtocol/aztec-packages: yellow-paper/docs/contracts/index.md

:::

Some additional discussion/comments on the message structure can be found in [The Republic](https://forum.aztec.network/t/the-republic-a-flexible-optional-governance-proposal-with-self-governed-portals/609/2#supporting-pending-messages-5).

Since any data that is moving from one chain to the other at some point will live on L1, it will be PUBLIC. While this is fine for L1 consumption (which is public in itself), we want to ensure that the L2 consumption can be private.
To support this, we use a nullifier scheme similar to what we are doing for all the other notes (**REFERENCE**). As part of the nullifier computation we then use the `secret` which hashes to the `secretHash`, this ensures that only actors with knowledge of `secret` will be able to see when it is spent on L2.

Any message that is consumed on one side MUST be moved to the other side. This is to ensure that the messages exist AND are only consumed once. The L1 contracts can handle one side, but the circuits must handle the other.

:::info Is `secretHash` required?

> We are using the `secretHash` to ensure that the user can spend the message privately with a generic nullifier computation. However, as the nullifier computation is almost entirely controlled by the app circuit (except the siloing) applications could be made to simply use a different nullifier computation and have it become part of the content. However, this reduces the developer burden and is quite easy to mess up. For those reasons we have decided to use the `secretHash` as part of the message.

Author: I think the siloing should be mentioned as part of the state writeup which did not exist as the time of writing this. But I can add another of the **REFERENCE** in there such that it is easy to find.

:::

### Inbox
When we say inbox, we are generally referring to the L1 contract that handles the L1 to L2 messages.

The inbox is logically a multi-set that builds messages based on the caller and user-provided content (multi-set meaning that repetition are allowed). While anyone can insert messages into the inbox, only the recipient state transitioner can consume messages from it (as specified by the version). When the state transitioner is consuming a message, it MUST insert it into the "L2 outbox" (message tree).

When a message is inserted into the inbox, the inbox **MUST** fill in the following fields:
- `L1Actor.actor`: The sender of the message (the caller), `msg.sender`
- `L1Actor.chainId`: The chainId of the L1 chain sending the message, `block.chainId`

We MUST populate these values in the inbox, since we cannot rely on the user providing anything meaningful. From the `L1ToL2Msg` we compute a hash of the message. This hash is what is moved by the state transitioner to the L2 outbox. 

> Since message from L1 to L2 can be inserted independently of the L2 block, the message transfer (insert into inbox move to outbox) are not synchronous as it is for L2 to L1. This means that the message can be inserted into the inbox, but not yet moved to the outbox. The message will then be moved to the outbox when the state transitioner is consuming the message as part of a block. Since the sequencers are responsible for the ordering of the messages, there is not a known time for this pickup to happen, it async. 

Author: Oh, it is async it should say, just to make the point clear. Will add the extra word.

This is done to ensure that the messages are not used to DOS the state transitioner. If the state transitioner was forced to pick up the messages in a specific order or at a fixed rate, it could be used to DOS the state transitioner by inserting a message just before a rollup goes through. 
While this can be addressed by having a queue of messages and let the sequencer specify the order, this require extra logic and might be difficult to price correctly. To keep this out of protocol, we simply allow the user to attach a fee to the message (see `fee` in `L1ToL2Msg` above). This way, the user can incentivize the sequencer to pick up the message faster.

> Since it is possible to land in a case where the sequencer will never pick up the message (e.g., if it is underpriced), the sender must be able to cancel the message. To ensure that this cancellation cannot happen under the feet of the sequencer we use a `deadline`, only after the deadline can it be cancelled. 

Author: Yes, that is what it means. You don't want to pick up stuff after the deadline.

- **Header Validation**:
    - The starting state of the block (derived from the header) MUST match the state stored in the contract
    - The global variables defined by the header MUST be valid:
        - The block number MUST be the next block number
        - The timestamp MUST:
            - be newer than the previous block inclusion
            - not be in the future
        - The version MUST be the same as the current version
        - The chainId MUST be the same as the current chainId
    - The ending state of the block (derived from the header) MUST *replace* the state stored in the contract
        - Requires ALL `MUST` constraints to be met
- **Proof validation**: The proof MUST be valid with the public inputs hash
    - A single public input hash MUST be computed from the block header, the commitment to the block content and the commitment to L1 to L2 messages.
- **State update**: The state root MUST be set to the ending state value
- **Inserting messages**: for messages that are inserted into the inboxes:
    - The `sender.actor` MUST be the caller

>     - The `(sender|recipient).chainId` MUST be the chainId of the L1 state transitioner

Author: Guess it kinda is. But will clarify

    - The `(sender|recipient).version` MUST be the version of the L1 state transitioner (the version of the L2)
    - The `content` MUST fit within a field element
    - For L1 to L2 messages:        
        - The `deadline` MUST be in the future, `> block.timestamp`
        - The `secretHash` MUST fit in a field element
        - The caller MAY append a `fee` to incentivize the sequencer to pick up the message
- **Message Cancellation**: To remove messages from the L1 inbox:
    - The message MUST exist in the inbox
    - The caller MUST be `sender.actor`
    - The current time (`block.timestamp`) MUST be larger than the `deadline`
    - The `fee` SHOULD be refunded to the caller
- **Moving messages**:
    - Moves MUST be atomic:
        - Any message that is inserted into an outbox MUST be consumed from the matching inbox
        - Any message that is consumed from an inbox MUST be inserted into the matching outbox

>     - Messages MUST be moved by the state transitioner matching the `version` of the message

Reviewer: > matching the `version` of the message
What do you mean by this?

- **Consuming messages**: for messages that are consumed from the outboxes:
    - L2 to L1 messages (on L1):
        - The consumer (caller) MUST match the `recipient.actor`
        - The consumer chainid MUST match the `recipient.chainId`
        - The consumer SHOULD check the `sender`
    - L1 to L2 messages (on L2):

>         - The consumer contract SHOULD check the `sender` details against the `portal` contract

Author: But it is not just the portal address, also the version. its the full sender.

        - The consumer contract SHOULD check that the `secret` is known to the caller
        - The consumer contract SHOULD check the `recipient` details against its own details
        - The consumer contract SHOULD emit a nullifier to ensure not double-spending
        - The consumer contract SHOULD check that the message exists in the state

:::info

> - We compute a single hash since each public input increase costs of the proof verification.

Reviewer: ```suggestion
- We compute a single hash since each public input increases the costs of proof verification.
```

        loop tx in txs

            loop msg in tx.l1ToL2Consume
                P2->>O2: Consume msg
                O2->>O2: Validate msg
                O2->>O2: Update state (nullify)
            end

            loop msg in tx.l2ToL1Msgs
                P2->>I2: Add msg
                I2->>I2: Populate msg values
                I2->>I2: Update state (insert)
            end
        end

        loop msg in L2 inbox 
            R2->>O2: Consume msg
            O2->>O2: Update state (delete)
        end

        loop msg in l1ToL2Msgs 
            R2->>O2: Insert msg
            O2->>O2: Update state (insert)
        end

        R2->>R: Block (Proof + Data)

        R->>R: Verify proof
        R->>R: Update State 

        R->>Reg: Where is the Inbox?
        Reg->>R: Here is the address

        R->>I: Consume l1ToL2Msgs from L1
        I->>I: Update state (delete)

        R->>Reg: Where is the Outbox?
        Reg->>R: Here is the address

        R->>O: Insert Messages from L2
        O->>O: Update state (insert)

    end

    P->>O: Consume a msg
    O->>O: Validate msg
    O->>O: Update state (delete)
```
We will walk briefly through the steps of the diagram above.

> 1. A portal contracts on L1 want to send a message for L2

Reviewer: ```suggestion
1. A portal contracts on L1 wants to send a message for L2
```

> 1. The L1 nbox populates the message with `sender` information

Reviewer: ```suggestion
1. The L1 inbox populates the message with `sender` information
```

1. The L1 inbox contract inserts the message into its storage

> 1. On the L2, as part of a rollup, a transaction tries to consume a message from the L2 outbox.

Reviewer: ```suggestion
1. On the L2, as part of an L2 block, a transaction tries to consume a message from the L2 outbox.
```

1. The L2 outbox ensures that the message is included, and that the caller is the recipient and knows the secret to spend. (This is practically done by the application circuit)
1. The nullifier of the message is emitted to privately spend the message (This is practically done by the application circuit)
1. The L2 contract wishes to send a message to L1
1. The L2 inbox populates the message with `sender` and `recipient` information
1. The L2 inbox inserts the message into its storage
1. The rollup circuit starts consuming the messages from the inbox
1. The L2 inbox deletes the messages from its storage

> 1. The rollup include messages from the L1 inbox that are to be inserted into the L2 outbox.

Reviewer: ```suggestion
1. The L2 block contains messages from the L1 inbox that are to be inserted into the L2 outbox.
```

1. The outbox state is updated to include the messages

> 1. The rollup block is submitted to L1 

Reviewer: ```suggestion
1. The L2 block is submitted to L1 
```

1. The state transitioner receives the block and verifies the proof + validate constraints on block.
1. The state transitioner updates it state to the ending state of the block
1. The state transitioner ask the registry for the L1 inbox address
1. The state transitioner retrieves the L1 inbox address
1. The state transitioner consumes the messages from the L1 inbox that was specified in the block. Note that they have logically been inserted into the L2 outbox, ensuring atomicity.
1. The L1 inbox updates it local state by deleting the messages that was consumed
1. The state transitioner ask the registry for the L1 outbox address
1. The state transitioner retrieves the L1 outbox address
1. The state transitioner inserts the messages into the L1 inbox that was specified in the block. Note that they have logically been consumed from the L2 outbox, ensuring atomicity.
1. The L1 outbox updates it local state by inserting the messages
1. The portal later consumes the message from the L1 outbox
1. The L1 outbox validates that the message exists and that the caller is the recipient
1. The L1 outbox updates it local state by deleting the message

:::info L2 inbox is not real

> As should be clear from above, the L1 inbox don't need to exist for itself, it keeps no state between blocks, as every message created in the block will also be consumed in the same block. 

Reviewer: ```suggestion
As should be clear from above, the L1 inbox doesn't need to exist for itself, it keeps no state between blocks, as every message created in the block will also be consumed in the same block. 
```
